### ISSUE TITLE
Not using all GPU memory

### ISSUE THREAD

**Issue by rterwedo** [gurnec/btcrecover#48]
Its a bitcoin core wallet, on ubuntu 14.04 on AWS.  No matter what, it will not go above ~38mb on each thread/gpu.  This is at full computer/in pwd.  Any ideas on settings?

Fri Apr 14 03:02:14 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 375.51                 Driver Version: 375.51                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GRID K520           On   | 0000:00:04.0     Off |                  N/A |
| N/A   30C    P0    42W / 125W |     40MiB /  4036MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GRID K520           On   | 0000:00:05.0     Off |                  N/A |
| N/A   33C    P0    42W / 125W |     40MiB /  4036MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GRID K520           On   | 0000:00:06.0     Off |                  N/A |
| N/A   31C    P0    41W / 125W |     40MiB /  4036MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      3190    C   python                                          38MiB |
|    1      3190    C   python                                          38MiB |
|    2      3190    C   python                                          38MiB |
+-----------------------------------------------------------------------------+

**Comment by gurnec** [comment]
Bitcoin Core wallets don't require much GPU RAM to search for passwords; they use a key derivation algorithm called PBKDF1 which isn't designed to use much RAM. In other words this is normal.

(Armory, on the other hand, benefits from a lot of GPU RAM.)

**Comment by rterwedo** [comment]
Hi - So what setup would be optimized for BTC core wallets if GPU ram is not needed?

Many GPUs? CPUs? Sys RAM?

**Comment by gurnec** [comment]
*Probably* many GPUs, but it depends. If you have a complicated token file or are using many command-line options or typos, many GPUs will only help up to a certain point. In other words, please don't buy 4 or 6 GPUs and expect it to be 4 or 6 times faster than 1 GPU - it's possible it *might* be that fast, but it's also possible that there could be other limiting factors.

You can run btcrecover with a token file and all the options you want, but no wallet by adding `--no-eta --wallet __null` to the command line. This will give you the max P/s rate you could possibly get with that token file/typos. Let's say it's 80kP/s.

A GTX 1080 will get *very roughly* about 25kP/s (but it might get more or less; it also depends on the specific wallet.dat file). This means that getting more than 3 1080s would be pointless - the CPU is then the limiting factor. Even getting three might get you less than the expected 75kP/s - it just doesn't scale perfectly linearly.

You can use PyPy to speed up that 80kP/s max number, but it's a little tricky. Take a look at [this comment starting with the paragraph "The second workaround is to use PyPy"](https://github.com/gurnec/btcrecover/issues/64#issuecomment-307216902).

 Finally, NVIDIAs seem to be the way to go right now, specifically the 10xx series.

**Comment by rterwedo** [comment]
Wow.  Well something was for sure off.  I was using AWS 8xGPU (4 Nvidia K520 Teslas - big beasts) and was getting 2500 pass/sec.  I filed this bug because I assumed it was being limited by the lack of RAM usage in the GPU.  It seems something else is way off instead...

I am looking for screens I may have took re: # of passwords a second it was doing.

**Comment by rterwedo** [comment]
Thanks for pypy recommendation, helps tons generating the list.

However actually trying them with the 4 GPUs I mentioned is showing around 2.500 password tries per sec.  Is this normal for a BTC core wallet?  I could not tell if you were talking of ~25k pwd /s in generation or trying mode.

**Comment by gurnec** [comment]
It looks like the K520s should be similar to GTX 690s. Here are rough performance numbers from some other GPUs:

 * GTX 560 Ti - 1 kP/s
 * R9 390 - 4.5 kP/s
 * GTX 1080 - 25 kP/s (reportedly, haven't tried it myself)

2.5 kP/s from 4x K520s definitely seems slow to me. What `--global-ws` and `--local-ws` are you using? Would you mind posting *just the last 12 characters* output from [running `extract-bitcoincore-mkey.py`](https://github.com/gurnec/btcrecover/blob/master/docs/Extract_Scripts.md#usage-for-bitcoin-classicxtcore)?

**Comment by rterwedo** [comment]
Re: --global-ws and --local-ws  ... it was over a month ago so I cannot recall.  I tried many different combos per your docs.  I never really got above 2.800 pwd / s (cracking).  Generating I am now seeing 10x that now with using pypy like you said which still seems low but is OK.

Re: last 12.  I cannot run that command on my mac, and cloud Ubuntu due to bssd issues (which I cannot remember how I solved before) BUT I think this is from a successful run 2 months ago 

[https://zerobin.net/?b00538b13b3fe9be#PVPAqjKqprNIfcP/CGf+4mds743mzSCh8j8W9r30z24=]

**Comment by gurnec** [comment]
> Re: --global-ws and --local-ws ... it was over a month ago so I cannot recall.

Fair enough. When I have some spare cycles, I'll try to reproduce this on a modern EC2 GPU instance.

> I think this is from a successful run 2 months ago

Thanks. Each wallet has a different "iteration" count, which affects how long it takes to try passwords. I was just trying to make sure that your wallet wasn't out of the ordinary -- the one you posted has a count of 105228 which is typical.

```python
>>> import struct
>>> struct.unpack('I', 'qAybAQCX57n+'.decode('base64')[1:5])
(105228,)
```

**Comment by rterwedo** [comment]
I can throw a few bucks your way to help sponsor if you want.  Zerobin me contact/payment info, I see a consulting email for a domain thats not up in your profile.

I was using 8xlarge SPOT in EU (or wherever was cheapest)

**Comment by rterwedo** [comment]
Hi @gurnec - looking to build 1 (or more if needed) rigs cost effectively to get to 50k-100k p/s on a btc core wallet.  Do you have some recommended hardware configs (seems gtx 1080s @25k p/s sound good?)  I dont need latest /greatest, just some ideas on GPU, CPU, RAM setups you have come across.  I am at 10s to 100s of billions of guesses I need to go over, with token files in the hundreds of gigs.  I can run token files on AWS or some other cloud service if a rig that is optimized for the testing the pwds is not optimized for token generation.  Any helpful pointers?

**Comment by rterwedo** [comment]
@gurnec - Assuming the password list is already generated, and with a --wallet __null getting 200k p/s, a real wallet of 105k SHA512, do you think CPU will be a limiting factor?  Even with a a couple 1080s, seems prob not going to hit theoretical CPU limit.  Correct assumption?

**Comment by nori19870522** [comment]
I am Japanese and do password analysis.
https://keychainx.io/ 

This is Robert's article.
https://news.bitstarz.com/bitstarz-news-interview-robert-rhodin-keychainx-ceo  

I may be able to find the password you lost. Do you need help with your login?

**Comment by btcexchanging** [comment]
if you haven't recovered your wallet, i can help you recover your wallet.
I have recovered most of the common wallets and recovery phrases for many users
you can reach me here at my telegram account:  @walletrecover
